appcelerator/titanium_mobile
# fix(ios): remove additional gc protection once proxy is remembered

**JIRA:** https://jira.appcelerator.org/browse/TIMOB-27207

**Optional Description:**
To protected JS objects after creation and while they are not referenced on the stack or in the JS object graph we introduced `applyGarbageCollectionSafeguard`. The GC protection is removed in https://github.com/appcelerator/titanium_mobile/blob/58c26e10d4641d36b7ac7f28fd61fd0ecb9d394e/iphone/TitaniumKit/TitaniumKit/Sources/API/TiBindingTiValue.m#L202 when a value crosses the native bridge. This works fine for statements like `const table = Ti.UI.createTableView()`.

However, this can cause issues for some internal values, like the `TiUITableViewSectionProxy` of a table view. Unless you explicitly access that section in your code (which will force it to cross the native bridge, e.g. `const section = table.data[0]`) it will remain GC protected. Since we already make sure that those proxies don't get GC'ed by using `rememberProxy`, this fix will remove the creation time GC safeguard in this method too.

**Testing steps**

- Use the provided example app in the ticket or copy & paste the following simplified example:
  ```js
  const win = Ti.UI.createWindow({
    layout: 'vertical'
  });
  const table = Ti.UI.createTableView({
    data: [ {title: 'Apples'}, {title: 'Bananas'}, {title: 'Carrots'}, {title: 'Potatoes'} ]
  });
  const btn = Ti.UI.createButton({
    title: 'Reload',
    top: 40
  });
  btn.addEventListener('click', () => {
    table.data = [ {title: 'Apples'}, {title: 'Bananas'}, {title: 'Carrots'}, {title: 'Potatoes'} ];
  });
  win.add(btn);
  win.add(table);
  win.open();
  ```
- Build and run the app. Open Instruments with the Leaks template, attach to the running app and filter for `tiui`.
- Click the Reload button multiple times
- `TiUITableViewRowProxy` and `TiUITableViewSectionProxy` entries show up with an increasing number of persistent objects.
- Wait a short time until the JS garbage collection starts. Most of the object count should move to Transient now.
